File: EX9/sort_time.py
import time
import random

array = [random.uniform(0, 1000) for _ in range(1000000)]

start_time = time.time()
sorted_array = sorted(array)
end_time = time.time()

print("Time taken to sort:", end_time - start_time, "seconds")

def sort(array): # Algoritmo quicksort gerado pelo copilot
    if len(array) <= 1:
        return array
    pivot = array[len(array) // 2]
    left = [x for x in array if x < pivot]
    middle = [x for x in array if x == pivot]
    right = [x for x in array if x > pivot]
    return sort(left) + middle + sort(right)


start_time_custom = time.time()
sorted_array_custom = sort(array)
end_time_custom = time.time()

print("Time taken to sort with custom sort:", end_time_custom - start_time_custom, "seconds")


array_2m = [random.uniform(0, 1000) for _ in range(2000000)]

start_time_2m = time.time()
sorted_array_2m = sorted(array_2m)
end_time_2m = time.time()

print("Time taken to sort 2 million elements:", end_time_2m - start_time_2m, "seconds")

start_time_custom_2m = time.time()
sorted_array_custom_2m = sort(array_2m)
end_time_custom_2m = time.time()

print("Time taken to sort 2 million elements with custom sort:", end_time_custom_2m - start_time_custom_2m, "seconds")

File: EX9/sort_time.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <chrono>

void custom_sort(std::vector<float>& arr, int low, int high) { // algoritmo quicksort gerado pelo copilot
    if (low < high) {
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                std::swap(arr[i], arr[j]);
            }
        }
        std::swap(arr[i + 1], arr[high]);
        custom_sort(arr, low, i);
        custom_sort(arr, i + 2, high);
    }
}

int main() {
    std::vector<float> arr(1000000);
    std::generate(arr.begin(), arr.end(), []() { return static_cast<float>(std::rand()) / (static_cast<float>(RAND_MAX / 1000.0)); });

    auto start = std::chrono::high_resolution_clock::now();

    std::sort(arr.begin(), arr.end());

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;

    std::cout << "Time taken to sort: " << elapsed.count() << " seconds" << std::endl;

    std::generate(arr.begin(), arr.end(), []() { return static_cast<float>(std::rand()) / (static_cast<float>(RAND_MAX / 1000.0)); });

    start = std::chrono::high_resolution_clock::now();

    custom_sort(arr, 0, arr.size() - 1);

    end = std::chrono::high_resolution_clock::now();
    elapsed = end - start;

    std::cout << "Time taken to custom sort: " << elapsed.count() << " seconds" << std::endl;

    arr.resize(2000000);
    std::generate(arr.begin(), arr.end(), []() { return static_cast<float>(std::rand()) / (static_cast<float>(RAND_MAX / 1000.0)); });

    start = std::chrono::high_resolution_clock::now();

    std::sort(arr.begin(), arr.end());

    end = std::chrono::high_resolution_clock::now();
    elapsed = end - start;

    std::cout << "Time taken to sort 2 million elements: " << elapsed.count() << " seconds" << std::endl;

    std::generate(arr.begin(), arr.end(), []() { return static_cast<float>(std::rand()) / (static_cast<float>(RAND_MAX / 1000.0)); });

    start = std::chrono::high_resolution_clock::now();

    custom_sort(arr, 0, arr.size() - 1);

    end = std::chrono::high_resolution_clock::now();
    elapsed = end - start;

    std::cout << "Time taken to custom sort 2 million elements: " << elapsed.count() << " seconds" << std::endl;

    return 0;
}


